Inc/C++Utilities/Arena.hpp

```diff
@@ -0,0 +1,179 @@
+        if (!usedIndexesSet[index]) {
+            return false; // Double free detected
+        }
+        ele->~T();
+        return release(ele);
+    }
+
+    size_t capacity() const { return S; }
+    size_t available() const { return freeIndexes.size(); }
+    size_t used() const { return S - freeIndexes.size(); }
+
+    /**
+     * @brief Iterator for traversing used elements in the arena.
+     */
+    class Iterator {
+    public:
+        Iterator(Arena* arena, size_t index) : arena(arena), index(index) {
+            // Skip to the first used element
+            while (index < S && !arena->usedIndexesSet[index]) {
+                ++index;
+            }
+            this->index = index;
+        }
+
+        T& operator*() { return arena->elements[index]; }
+        T* operator->() { return &arena->elements[index]; }
+
+        Iterator& operator++() {
+            do {
+                ++index;
+            } while (index < S && !arena->usedIndexesSet[index]);
+            return *this;
+        }
+
+        bool operator!=(const Iterator& other) const {
+            return index != other.index;
+        }
+
+    private:
+        Arena* arena;
+        size_t index;
+    };
+
+    /**
+     * @brief Const iterator for traversing used elements in the arena.
+     */
+    class ConstIterator {
+    public:
+        ConstIterator(const Arena* arena, size_t index) : arena(arena), index(index) {
+            // Skip to the first used element
+            while (index < S && !arena->usedIndexesSet[index]) {
+                ++index;
+            }
+            this->index = index;
+        }
+
+        const T& operator*() const { return arena->elements[index]; }
+        const T* operator->() const { return &arena->elements[index]; }
+
+        ConstIterator& operator++() {
+            do {
+                ++index;
+            } while (index < S && !arena->usedIndexesSet[index]);
+            return *this;
+        }
+
+        bool operator!=(const ConstIterator& other) const {
+            return index != other.index;
+        }
+
+    private:
+        const Arena* arena;
+        size_t index;
+    };
+
+    Iterator begin() { return Iterator(this, 0); }
+    Iterator end() { return Iterator(this, S); }
+    ConstIterator begin() const { return ConstIterator(this, 0); }
+    ConstIterator end() const { return ConstIterator(this, S); }
+    ConstIterator cbegin() const { return ConstIterator(this, 0); }
+    ConstIterator cend() const { return ConstIterator(this, S); }
+
+    Arena() {
+        // Push indices in reverse order so index 0 is allocated first
+        for (int i = S - 1; i >= 0; --i) {
+            freeIndexes.push(i);
+        }
+    }
+    Arena(const Arena&) = delete;
+    Arena& operator=(const Arena&) = delete;
+    Arena(Arena&& other) noexcept = delete;
+    Arena& operator=(Arena&& other) = delete;
+
+   private:
+    T elements[S];
+    Stack<S> freeIndexes;
+    bool usedIndexesSet[S]{false};
```

Comment:
Could use a bitmap instead of a bool array then getting the first used one is as easy as ctz(used_idx_bm), loop by clearing LSB and ctz again. Just make sure you static assert that S is below the number of bits of the bitmap